primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,124 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameters are encoded as (u8, u64) tuples, where u8 is a tag defining a type of
+//! parameter, and u64 is a parameter value. Parameter set and value structure are opaque
+//! and depend on execution environment itself. Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use parity_scale_codec::{Decode, Encode};
+use parity_util_mem::MallocSizeOf;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+	/// Legacy instantiation strategy
+	Legacy,
+}
+
+/// A single executor parameter
+#[non_exhaustive]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutorParam {
+	/// General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
+	/// Version of the set of execution parameters, unique in the scope of execution environment type
+	Version(u32),
+
+	/// Parameters setting the executuion environment semantics:
+	/// Number of extra heap pages
+	ExtraHeapPages(u64),
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit, in units
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+	/// Executor instantiation strategy
+	InstantiationStrategy(ExecInstantiationStrategy),
+	/// `true` if compiler should perform NaNs canonicalization
+	CanonicalizeNaNs(bool),
+	/// `true` if parallel compilation is allowed, single thread is used otherwise
+	ParallelCompilation(bool),
+}
+
+/// Deterministically serialized execution environment semantics
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub struct ExecutorParams(Vec<ExecutorParam>);
+
+impl ExecutorParams {
+	/// Returns globally unique version of execution environment parameter set, if present. Otherwise, returns 0.
+	pub fn version(&self) -> u64 {
+		if self.0.len() < 2 {
+			return 0
+		}
+		let env = if let ExecutorParam::Environment(env) = self.0[0] { env } else { return 0 };
+		let ver = if let ExecutorParam::Version(ver) = self.0[1] { ver } else { return 0 };
```

Comment:
nit: Just thinking here, but is it possible to use the type system to encode some of these assumptions about the layout? i.e. The params can be either empty, or else they *must* have an env and a ver, so why not have it be an `Option` of a struct that contains env, var, and additional params? The additional params can be a vec that can grow without bounds and the whole thing can be `repr(C)` if needed.

Reply:
Well spotted, but I still didn't want to make it less homogeneous so I've joined environment type and version in a single enum element.

Reply:
Nice one. Resolving.